Camel-freemarker component

carlosthe19916 opened the pull request:
Issue: https://github.com/apache/camel-quarkus/issues/223

Added camel-freemarker component.
Changes:
- The same test used for apache-camel-freemarker
- We can add the list of locations where the templates can be found in the classpath using: `quarkus.camel.freemarker.locations=myFolder1,myFolder2`

Looking forward to have some feedback.

oscerd commented on extensions/freemarker/deployment/src/main/java/org/apache/camel/quarkus/component/freemarker/deployment/FreemarkerNativeImageProcessor.java:
```diff
@@ -0,0 +1,120 @@
+package org.apache.camel.quarkus.component.freemarker.deployment;
```
Please add the ASF license header.

carlosthe19916 replied:
@oscerd Added, thanks

ppalaga commented on extensions/freemarker/deployment/src/main/java/org/apache/camel/quarkus/component/freemarker/deployment/FreemarkerProcessor.java:
```diff
@@ -0,0 +1,54 @@
+    @BuildStep
+    void registerForReflection(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, true,
+                Exchange.class,
```
Is Exchange really required here?

carlosthe19916 replied:
Yes, adding Exchange for reflection is needed. According to this documentation: [freemarker context](https://camel.apache.org/components/latest/freemarker-component.html#_freemarker_context), it should be possible to use `exchange.properties` in the `.ftl` templates just like this example: https://github.com/carlosthe19916/camel-quarkus/blob/7d45a783e5b27d341b6a07df49b34087f636dad3/integration-tests/freemarker/src/main/resources/org/apache/camel/component/freemarker/example.ftl#L19. Without using reflection for Exchange it won't work.

ppalaga commented on extensions/freemarker/deployment/src/main/java/org/apache/camel/quarkus/component/freemarker/deployment/FreemarkerProcessor.java:
```diff
@@ -0,0 +1,54 @@
+
+    @BuildStep
+    void registerForReflection(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, true,
```
Wouldn't `new ReflectiveClassBuildItem(true, false, ...` be enough?

carlosthe19916 replied:
You're right, I'll update that, Thanks :)

ppalaga commented on extensions/freemarker/runtime/pom.xml:
```diff
@@ -0,0 +1,82 @@
+    <name>Camel Quarkus :: Freemarker :: Runtime</name>
+
+    <properties>
+        <firstVersion>0.4.0</firstVersion>
```
```suggestion
        <firstVersion>1.0.0-M5</firstVersion>
```

and re-run `mvn process-resources -Pformat -Denforce=false`

carlosthe19916 replied:
done

ppalaga commented on integration-tests/freemarker/src/main/resources/application.properties:
```diff
@@ -0,0 +1,20 @@
+#
+# Quarkus
+#
+quarkus.camel.freemarker.locations = org/apache/camel/component/freemarker
```
Given that `org/apache/camel/component/freemarker` will be scanned recursivelly, I wonder whether all the class files are going to get included as native image resources? I would be interesting to compare the size of the native image using this locations value and using a new path that does not contain any class files.

carlosthe19916 replied:
Only files with the extension `.ftl` will be included. You can see there is a filter for that [here](https://github.com/carlosthe19916/camel-quarkus/blob/7d45a783e5b27d341b6a07df49b34087f636dad3/extensions/freemarker/deployment/src/main/java/org/apache/camel/quarkus/component/freemarker/deployment/FreemarkerNativeImageProcessor.java#L119)... I've also created 2 native images, the first one using `org/apache/camel/component/freemarker` and the second one using `myfreemarkerfolder` and the sizes are exactly the same (45.8 MB)

ppalaga replied:
> Only files with the extension `.ftl` will be included.

Thanks for the info, I have not noticed that and I must say I do not think we should be selecting only `*.ftl` files. I commonly used other or no extensions for FreeMarker template files in the past. I think many people do that because because of syntax highlighting in editors. There is even a couple of examples in this repo: https://github.com/apache/camel-quarkus/tree/master/tooling/create-extension-templates

I think we should either include whole subtrees or we should have path includes and excludes like we have in https://github.com/apache/camel-quarkus/blob/master/extensions/core/runtime/src/main/java/org/apache/camel/quarkus/core/CamelConfig.java#L74-L100 WDYT?

carlosthe19916 replied:
Good point @ppalaga . Files with different extensions should still work. I agree and we should include the whole subtree but just to avoid adding files we don't want to add we can use `includePatterns` and `excludePatterns` as you suggested. Let me add it

carlosthe19916 replied:
@ppalaga sorry for being so slow in this. I've added `includePatterns` and `excludePatterns`

ppalaga commented:
Could you please rebase and squash your commits? It would make it easier to review locally.

carlosthe19916 commented on extensions/freemarker/deployment/src/main/java/org/apache/camel/quarkus/component/freemarker/deployment/FreemarkerNativeImageProcessor.java:
```diff
@@ -0,0 +1,136 @@
+    void build(
+            BuildProducer<NativeImageResourceBuildItem> resourceProducer,
+            CamelFreemarkerConfig camelFreemarkerConfig) throws IOException, URISyntaxException {
+        registerNativeImageResources(resourceProducer, camelFreemarkerConfig);
```
Thanks for the good advice, I'll do it inline just like you suggested :)

carlosthe19916 replied:
done

carlosthe19916 commented on extensions/freemarker/deployment/src/main/java/org/apache/camel/quarkus/component/freemarker/deployment/FreemarkerNativeImageProcessor.java:
```diff
@@ -0,0 +1,136 @@
+            for (String location : locations) {
+                // Strip any 'classpath:' protocol prefixes because they are assumed
+                // but not recognized by ClassLoader.getResources()
+                if (location != null && location.startsWith(CLASSPATH_APPLICATION_MIGRATIONS_PROTOCOL + ':')) {
```
my mistake, it shouldn't be possible to have null values since it comes from `CamelFreemarkerConfig`

ppalaga replied:
Hm..., I wonder how the MicroProfile config impl used by Quarkus would parse a double comma value like this

```
quarkus.camel.freemarker.locations = my/path1,,my/path2
```

carlosthe19916 replied:
@ppalaga good point. I've verified that `my/path1,,my/path2` will cause just 2 locations but `my/path1, ,my/path2` (notice the blank space) will generate 3 locations and generate an exception... I've added a basic filter to validate it https://github.com/carlosthe19916/camel-quarkus/blob/7c027254d03024ec7d64aeffd040ae05da4677ff/extensions/freemarker/deployment/src/main/java/org/apache/camel/quarkus/component/freemarker/deployment/FreemarkerNativeImageProcessor.java#L66

I also added this case to the test here https://github.com/carlosthe19916/camel-quarkus/blob/7c027254d03024ec7d64aeffd040ae05da4677ff/integration-tests/freemarker/src/main/resources/application.properties#L20

ppalaga replied:
Thanks for the research, I have learned something.

carlosthe19916 commented on extensions/freemarker/deployment/src/main/java/org/apache/camel/quarkus/component/freemarker/deployment/FreemarkerNativeImageProcessor.java:
```diff
@@ -0,0 +1,136 @@
+        try {
+            List<String> locations = new ArrayList<>(camelFreemarkerConfig.locations);
+            if (locations.isEmpty()) {
+                locations.add("freemarker/templates");
```
I've added the default location `freemarker/templates` to CamelFreemarkerConfig and also added some docs in `docs/modules/ROOT/pages/extensions/freemarker.adoc`

carlosthe19916 commented:
@ppalaga I've also added a new Config property called `classModels` which should be used to pass all Java Classes that will be marked for reflection. Freemarker uses POJOs for processing the templates but if those POJOs were not marked for reflection using Quarkus, then an exception will happen... Because of this change, I've removed this line https://github.com/apache/camel-quarkus/pull/835/commits/0e2d854d776a9e656b2bfafc40826ee330710b1b#diff-d5f7db497629630202e093f916c76b9eL65 WDYT?

ppalaga commented on docs/modules/ROOT/pages/extensions/freemarker.adoc:
```diff
@@ -0,0 +1,53 @@
+|===
+|Configuration property |Type |Default
+
+|`quarkus.camel.freemarker.locations`
```
Could we perhaps remove the locations option now that we have includes and excludes? The includes option seems to offer a superset of `locations` functionality. The default can be changed to `freemarker/templates/**` and moved to the includes option too. WDYT?

ppalaga commented on docs/modules/ROOT/pages/extensions/freemarker.adoc:
```diff
@@ -0,0 +1,53 @@
+
+|`quarkus.camel.freemarker.include-patterns`
+
+ Used for inclusive filtering scanning of Freemarker templates. By default all tree files inside 'locations' will be included.
```
Plase add:

* Where we scan for the templates (classpath, filesystem)
* Are `classpath:` and `filesystem:` prefixes supported? Which one is the default? (I do not insist on supporting `filesystem:`) We have other extensions documenting similar features properly, just grep for `classpath:` in the source tree.

ppalaga commented:
It would be cool if the extension could do most/all of the following tasks at build time:

* Create a `freemarker.template.Configuration`
* Parse/warmup the templates.

That would make it truly quarkyfied!

ppalaga commented:
The tests were adapted in https://github.com/apache/camel-quarkus/pull/2028 and I'd like to implement the native support using the Quarkiverse Freemarker extension. Hence closing this one.
